File: src/main/java/algo/interval/CarPooling_1094.java
package algo.interval;

import java.util.TreeMap;

public class CarPooling_1094 {
  public static boolean carPooling(int[][] trips, int capacity) {
    var trip = new TreeMap<Integer, Integer>();  // time -> passengers (positive for taking passengers, negative for dropping)
    for (var t : trips) {
      trip.put(t[1], trip.getOrDefault(t[1], 0) + t[0]);  // take passenger
      trip.put(t[2], trip.getOrDefault(t[2], 0) - t[0]);  // drop passenger
    }

    var passengers = 0;
    for (var entry : trip.entrySet()) {
      passengers += entry.getValue();
      // at ant time, passengers should not exceed capacity
      if (passengers > capacity) {
        return false;
      }
    }
    return true;
  }

  public static void main(String[] args) {
    System.out.println(carPooling(new int[][]{{2,1,5},{3,3,7}}, 4));
    System.out.println(carPooling(new int[][]{{2,1,5},{3,3,7}}, 5));
  }
}

File: src/main/kotlin/algo/interval/CarPooling_1094.kt
package algo.interval

class KtCarPooling_1094 {
    fun carPooling(trips: Array<IntArray>, capacity: Int): Boolean {
        val trip: MutableMap<Int, Int> = mutableMapOf()
        trips.forEach { t ->
            var p = trip.getOrPut(t[1]) { 0 }
            trip[t[1]] = p + t[0]  // take passengers
            p = trip.getOrPut(t[2]) { 0 }
            trip[t[2]] = p - t[0]  // drop passengers
        }
        val sorted = trip.toSortedMap()
        var load = 0
        sorted.forEach { (_, v) ->
            load += v
            if (load > capacity) {
                return false
            }
        }
        return true
    }
}

fun main() {
    val pooling = KtCarPooling_1094()
//    println(pooling.carPooling(arrayOf(intArrayOf(2,1,5), intArrayOf(3,3,7)), 4))
    println(pooling.carPooling(arrayOf(intArrayOf(4,5,6), intArrayOf(6,4,7), intArrayOf(4,3,5), intArrayOf(2,3,5)), 13))
}